src/server/worker/api_server.go

```diff
@@ -200,8 +206,15 @@ func (logger *taggedLogger) Write(p []byte) (_ int, retErr error) {
 }
 
 func (logger *taggedLogger) Close() (*pfs.Object, int64, error) {
+	close(logger.msgCh)
 	if logger.putObjClient != nil {
+		if err := logger.eg.Wait(); err != nil {
+			return nil, 0, err
+		}
 		object, err := logger.putObjClient.CloseAndRecv()
+		// we set putObjClient to nil so that future calls to Logf won't send
+		// msg down logger.msgCh as we've just closed that channel.
+		logger.putObjClient = nil
```

Comment:
The last thing that runs in `Process` is a final log statement which records that `Process` has exited as well as the return value and error if there is one. This is a bit of an awkward situation since we don't know what the return value is or whether the function will error until we finish writing logs to the stats repo and serialize that repo. Unfortunately that means that the final log message can't be included in the stats repo.